src/v/cloud_storage/partition_manifest.cc

```diff
@@ -351,38 +351,44 @@ partition_manifest::full_log_start_kafka_offset() const {
           _archive_start_offset,
           _start_offset);
         return _archive_start_offset - _archive_start_offset_delta;
-    } else if (_start_offset != model::offset{}) {
-        return compute_start_kafka_offset_local();
     }
-    return std::nullopt;
+
+    return get_start_kafka_offset();
 }
 
 std::optional<kafka::offset>
 partition_manifest::get_start_kafka_offset() const {
-    if (_start_offset == model::offset{}) {
+    if (_start_offset == model::offset{} || !_cached_start_offset_delta) {
         return std::nullopt;
     }
-    return compute_start_kafka_offset_local();
+
+    return _start_offset - _cached_start_offset_delta.value();
 }
 
-std::optional<kafka::offset>
-partition_manifest::compute_start_kafka_offset_local() const {
-    std::optional<kafka::offset> local_start_offset;
+void partition_manifest::set_start_offset(model::offset start_offset) {
+    _start_offset = start_offset;
+    // `_cached_start_offset_delta` needs to be invalidated every
+    // time the start offset changes.
+    _cached_start_offset_delta = std::nullopt;
+}
```

Comment:
nit: I'm not sure, but it might help clean things up by making this take an offset_delta as an argument as well. Some call sites might need to be reordered, but I think it'd work out to something relatively tidy

Reply:
I'm thinking we could implement with the following invariant in mind: if there is a _start_offset, there must be a _cached_start_offset_delta.

The challenge is a mix of the cases where:
- `_start_offset` is {}
- `_segments` is empty

I think these cases will ultimately be the same, returning nullopt for a kafka offset.

Reply:
Yeah, we'd have to change when `_start_offset` is set in a few functions like `do_update` where the segment set is added after `_start_offset` is set. But that doesn't look like it would require much work.

My biggest concern is [here](https://github.com/redpanda-data/redpanda/pull/11964/files#diff-9284b0b8aebba5228faad2a6c2f98bb00d6078e4cf83a5774f51d8d0b915cfebL376) where I guess technically a start offset could be after the segments in `_segments`. In which case a `_cached_start_offset_delta` wouldn't exist. Though I don't believe that could happen. And the kafka offset would still be a nullopt in the current implementation. 

Will push a SQUASH commit that makes the change assuming the invariant.